packages/firebase_dynamic_links/ios/Classes/FLTFirebaseDynamicLinksPlugin.m

```diff
@@ -118,63 +118,57 @@ - (NSMutableDictionary *)getInitialLink {
+            [_channel invokeMethod:@"onLinkError"
+                         arguments:getDictionaryFromFlutterError(flutterError)];
+        } else {
+            NSMutableDictionary *dictionary = getDictionaryFromDynamicLink(dynamicLink);
+            [_channel invokeMethod:@"onLinkSuccess" arguments:dictionary];
+        }
+    } else {
+        if (error) {
+            _flutterError = getFlutterError(error);
+        } else {
+            _initialLink = dynamicLink;
+        }
+    }
 }
 
 - (BOOL)checkForDynamicLink:(NSURL *)url {
   FIRDynamicLink *dynamicLink = [[FIRDynamicLinks dynamicLinks] dynamicLinkFromCustomSchemeURL:url];
   if (dynamicLink) {
-    if (dynamicLink.url) _initialLink = dynamicLink;
+    [self onDeepLinkResult: dynamicLink error: nil];
     return YES;
   }
   return NO;
 }
 
-- (BOOL)onLink:(NSUserActivity *)userActivity {
-  BOOL handled = [[FIRDynamicLinks dynamicLinks]
-      handleUniversalLink:userActivity.webpageURL
-               completion:^(FIRDynamicLink *_Nullable dynamicLink, NSError *_Nullable error) {
-                 if (error) {
-                   FlutterError *flutterError = getFlutterError(error);
-                   [self.channel invokeMethod:@"onLinkError"
-                                    arguments:getDictionaryFromFlutterError(flutterError)];
-                 } else {
-                   NSMutableDictionary *dictionary = getDictionaryFromDynamicLink(dynamicLink);
-                   [self.channel invokeMethod:@"onLinkSuccess" arguments:dictionary];
-                 }
-               }];
-  return handled;
+- (BOOL)application:(UIApplication *)application
+            openURL:(NSURL *)url
+            options:(NSDictionary<UIApplicationOpenURLOptionsKey, id> *)options {
+  return [self checkForDynamicLink:url];
 }
 
-- (BOOL)onInitialLink:(NSUserActivity *)userActivity {
-  BOOL handled = [[FIRDynamicLinks dynamicLinks]
-      handleUniversalLink:userActivity.webpageURL
-               completion:^(FIRDynamicLink *_Nullable dynamicLink, NSError *_Nullable error) {
-                 if (error) {
-                   self.flutterError = getFlutterError(error);
-                 }
-                 self.initialLink = dynamicLink;
-               }];
-  return handled;
+- (BOOL)application:(UIApplication *)application
+              openURL:(NSURL *)url
+    sourceApplication:(NSString *)sourceApplication
+           annotation:(id)annotation {
+  return [self checkForDynamicLink:url];
 }
 
 - (BOOL)application:(UIApplication *)application
-    continueUserActivity:(NSUserActivity *)userActivity
-      restorationHandler:(void (^)(NSArray *))restorationHandler {
-  if (_initiated) {
-    return [self onLink:userActivity];
-  }
-  return [self onInitialLink:userActivity];
+continueUserActivity:(NSUserActivity *)userActivity
+ restorationHandler:(void (^)(NSArray *))restorationHandler {
+    BOOL handled = [[FIRDynamicLinks dynamicLinks]
```

Comment:
info: previously `_initiated` were checked before calling async function `[[FIRDynamicLinks dynamicLinks] handleUniversalLink:userActivity`. So previously if `getInitialLink` is called after the method is called but before it returns data. The link is lost.
After the change, the `_initiated` is checked after the callback returns.